pkg/tracker/apitracker.go

```diff
@@ -0,0 +1,285 @@
+/*
+Copyright © 2023 SUSE LLC
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tracker
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/Masterminds/log-go"
+	"github.com/containers/gvisor-tap-vsock/pkg/types"
+	"github.com/docker/go-connections/nat"
+)
+
+const (
+	// The Gateway IP address that is statically reserved
+	// by DHCP and will not change. It is used to initialize
+	// the NewAPITracker.
+	GatewayBaseURL = "http://192.168.127.1:80"
+	// Tap device (eth0) IP which is also allocated to the host-switch
+	// it is statically reserved by DHCP.
+	hostSwitchIP = "192.168.127.2"
+	exposeAPI    = "/services/forwarder/expose"
+	unexposeAPI  = "/services/forwarder/unexpose"
+)
+
+var (
+	ErrAPI         = errors.New("error from API")
+	ErrExposeAPI   = fmt.Errorf("error from %s API", exposeAPI)
+	ErrUnexposeAPI = fmt.Errorf("error from %s API", unexposeAPI)
+	ErrInvalidIPv4 = errors.New("not an IPv4 address")
+)
+
+// APITracker keeps track of the port mappings and calls the
+// corresponding API endpoints that is responsible for exposing
+// and unexposing the ports on the host. This should only be used when
+// the Rancher Desktop networking is enabled and the privileged service is disabled.
+type APITracker struct {
+	baseURL     string
+	httpClient  http.Client
+	portStorage *portStorage
+	*ListenerTracker
+}
+
+// NewAPITracker creates a new instace of a API Tracker.
+func NewAPITracker(baseURL string) *APITracker {
+	return &APITracker{
+		baseURL:         baseURL,
+		httpClient:      *http.DefaultClient,
+		portStorage:     newPortStorage(),
+		ListenerTracker: NewListenerTracker(),
+	}
+}
+
+// Add a container ID and port mapping to the tracker and calls the
+// /services/forwarder/expose endpoint to forward the port mappings.
+func (a *APITracker) Add(containerID string, portMap nat.PortMap) error {
```

Comment:
Trying to reword the documentation comment to see if I'm understanding the point of this method… (No need to change your comment, I'm just trying to check my understanding):

```
// Add the port mappings of a container to the port tracker to expose
// the ports on the VM host in order to keep up the illusion that the
// container engine is running directly instead of in a VM.
```

For an actual constructive comment: it might be good to refer to `Tracker.Add` here.